NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,587 @@
+                    # option to display named entities only
+                    if show_what==['NAMED ENTITIES']:
+                        ent_code = self.show_entities(doc, selTokens, inc_ent)
+                        my_code_list.insert(1,ent_code)
+                    
+                    # option to display speaker and/or quotes and/or named entities
+                    elif key in show_what:
+                        if 'NAMED ENTITIES' in show_what:
+                            ent_code = self.show_entities(doc, selTokens, inc_ent)
+                            my_code_list.insert(1,ent_code)
+                        
+                        start_token, end_token = selTokens[0], selTokens[-1] 
+                        span_code = "Span(doc, {}, {}, '{}'),".format(start_token, end_token+1, key) 
+                        my_code_list.insert(1,span_code)
+                    
+        # combine all codes
+        my_code = ''.join(my_code_list)
+    
+        # execute the code
+        exec(my_code)
+        
+        # display the preview in this notebook
+        displacy.render(doc, style='span', options=options, jupyter=True)
+        self.html = displacy.render(doc, style='span', options=options, jupyter=False, page=True)
+        
+        
+    def top_entities(self, text_id, which_ent='speaker_entities',top_n=5):
+        '''
+        Display top n named entities inside the text using displaCy
+
+        Args:
+            text_id: the text_id of the text you wish to display
+            which_ent: option to display named entities in speakers ('speaker_entities') 
+                       or quotes ('quote_entities')
+            top_n: the number of entities to display
+        '''
+        # get the top entities
+        most_ent = self.quotes_df[self.quotes_df['text_id']==text_id][which_ent].tolist()
+        most_ent = list(filter(None,most_ent))
+        most_ent = [ent for most in most_ent for ent in most]
+        most_ent = Counter([ent_name for ent_name, ent_label in most_ent])
+        top_ent = dict(most_ent.most_common()[:top_n])
+        
+        # visualize the top entities
+        bar_colors = {'speaker_entities':'#2eb82e',
+                      'quote_entities':'#008ae6'}
+        plt.figure(figsize=(10, 2.5))
+        plt.bar(top_ent.keys(), top_ent.values(), color=bar_colors[which_ent])
+        plt.yticks(range(0, most_ent[max(most_ent, key=most_ent.get)]+1, 1))
+        plt.title('Top {} {} in {}'.format(min(top_n,len(top_ent.keys())),which_ent,text_id))
+        plt.show()
+        
+
+    def analyse_quotes(self, inc_ent):
+        '''
+        Interactive tool to display and analyse speakers, quotes and named entities inside the text
+
+        Args:
+            inc_ent: a list containing the named entities to be extracted from the text, 
+                     e.g., ['ORG','PERSON','GPE','NORP','FAC','LOC']
+        '''
+        # widget for selecting text_id
+        enter_text = widgets.HTML(
```

Comment:
Let's try to use combobox for file input, and query on all filenames deduplicated by the unique file id?
https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#Combobox

Reply:
I just see your comment. I played around with auto-complete before, but wasn't very clean, so decided to go back to selection. will look at the combo button tomorrow.